### DISCUSSION TITLE
Override Application Resources

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#8616] [Q&A]
I'm looking to customize FluentAvalonia styles. For example, editing the background of buttons. I managed to restore the original colors of buttons by adding `Application.Resources` **AFTER** `Application.Styles`. Is there any way to make that generic so that I can re-use it in various projects?

```xaml
<Application.Resources>
    <StaticResource x:Key="ButtonBackground" ResourceKey="AccentFillColorDefaultBrush" />
</Application.Resources>
```

If I want to change the theme or accent color at runtime, however, the resources won't respect that. How can I adapt it so that the new theme/accent are applied?

If I have to edit the application resources via code, how do I do that?

**Comment by maxkatz6** [comment]
> buttons by adding Application.Resources AFTER Application.Styles

Order shouldn't matter.

> Is there any way to make that generic so that I can re-use it in various projects?

Not sure if I understood, but you can create a xaml file with ResourceDictionary inside with all of your reusable resources. And reference it with `<ResourceInclude>`.

> If I want to change the theme or accent color at runtime, however, the resources won't respect that

DynamicResource sees resources changes, when resource file is detached or attached (in case of ResourceInclude) or when Resource dictionary changed.

> If I have to edit the application resources via code, how do I do that?

control.Resources["Key"] = value; // where control can be Application

or `control.Resources.MergedDictionaries.Add()/Remove()`

More information here https://docs.avaloniaui.net/docs/styling/resources

**Reply by mysteryx93** [reply]
If I put it before, I get: Static resource 'AccentFillColorDefaultBrush' not found.

If I put it in a ResourceDictionary, I get the same problem.

These styles are defined by

```xaml
<Application.Styles>
    <sty:FluentAvaloniaTheme />
</Application.Styles>
```

To set the style at runtime, I need the IBrush instead of the string, how do I get it?

    Avalonia.Application.Current!.Resources["ButtonBackground"] = "ButtonBackground";

**Reply by maxkatz6** [reply]
> If I put it before, I get: Static resource 'AccentFillColorDefaultBrush' not found.

Yeah, it depends where you use it. I thought you are overriding existing style (ButtonBackground) in your Resoruces/ResourceDictionary. 
Button default control theme uses dynamic resource ButtonBackground, and it doesn't matter when to be overridden.
But if you have your own AccentFillColorDefaultBrush, and you want to statically reference it, this resource should be already included in the resources before `<StaticResource>` is calculated.

**Reply by maxkatz6** [reply]
> To set the style at runtime, I need the IBrush instead of the string, how do I get it?

control.TryFindResource (or Application.Current.TryFindResource)

**Reply by timunie** [reply]
For example: 

`Avalonia.Application.Current!.Resources["ButtonBackground"] = new SolidColorBrush(MyColor);`

**Reply by mysteryx93** [reply]
So is part of the problem that FluentAvalonia relies on StaticResources instead of DynamicResources or what? Might have to look into the source code to see how it switches from Light to Dark modes and recalculate accent tones and resources?

**Reply by timunie** [reply]
Well FluentAvalonia can be used to switch the accents quite easily. They have a themeProvider

```c#
// Get the service needed to change theme
FluentAvaloniaTheme? themeProvider => AvaloniaLocator.Current.GetService<FluentAvaloniaTheme>();

// Set your Accent
themeProvider.CustomAccentColor = MyAccentColor;
```

Happy coding
Tim

**Reply by mysteryx93** [reply]
Yes that's what I do to test whether the styles adapt. I found that if I set the accent color via code, the resource I set above does NOT take my custom accent color because it gets calculated before the custom theme gets applied. If I set it as a declarative argument, then it gets applied -- but won't adapt if I change it afterwards.

**Reply by mysteryx93** [reply]
Resources are defined here
https://github.com/amwx/FluentAvalonia/blob/master/FluentAvalonia/Styling/StylesV2/DarkResources.axaml#L155

`AccentFillColorDefaultBrush` and styles related to accent color are DynamicResource; others are StaticResource. So why isn't it adapting?

**Reply by timunie** [reply]
In that case I think you need to check the code there or file an issue there. Also it is helpful to see a sample of not working code for them, so consider to create a minimal sample.

**Reply by maxkatz6** [reply]
@mysteryx93 a lot of going on in this discussion... What is not adapting?
Just to clarify, `<StaticResource ResourceKey />` or `{StaticResource Key}` are executed and calculated at a point of time, when they added to the resource dictionary (at least before [this PR](https://github.com/AvaloniaUI/Avalonia/pull/8422)).
DynamicResource will track updates if resource with same name is replaced or whole resource dictionary was detached or attached.

**Reply by mysteryx93** [reply]
Here I'm referring `AccentFillColorDefaultBrush` defined as a DynamicResource in `DarkResources.axaml`. What happens if `DarkResources.axaml` gets removed altogether and replaced by "LightResources.axaml"?

Perhaps I should create an empty project to try to do just the base feature I'm trying to do.

**Reply by mysteryx93** [reply]
Actually in a simple project, it's adapting as expected. I'll try to do what I want in the basic project and update if I run into more issues.

**Reply by maxkatz6** [reply]
> What happens if DarkResources.axaml gets removed altogether and replaced by "LightResources.axaml"?

If AccentFillColorDefaultBrush  is used with DynamicResource, it will work.
If it will be used with StaticResource, it will work only if dictionary was replaced before StaticResource was instantiated.

**Reply by mysteryx93** [reply]
hum... this works

```xaml
<Application.Styles>
    <styling:FluentAvaloniaTheme />
<Application.Styles>
<Application.Resources>
    <SolidColorBrush x:Key="ButtonBackgroundApp" Color="{DynamicResource SystemAccentColorLight2}"/>
    <StaticResource x:Key="ButtonBackground" ResourceKey="ButtonBackgroundApp" />
</Application.Resources>
```

but this doesn't. Why?

```
<Application.Styles>
    <styling:FluentAvaloniaTheme />
    <Style Selector="Button /template/ ContentPresenter#PART_ContentPresenter">
        <Setter Property="Background" Value="{DynamicResource ButtonBackgroundApp}" />
    </Style>
<Application.Styles>
<Application.Resources>
    <SolidColorBrush x:Key="ButtonBackgroundApp" Color="{DynamicResource SystemAccentColorLight2}"/>
</Application.Resources>
```

I want to take the "accent" style and apply it by default with 35% background opacity; it's quite a lot of work to translate the code and find the proper references; but it seems to work.

**Comment by mysteryx93** [comment]
I've succeeded in setting the default style as "accent" with 35% background opacity! With gradient window background. Using a Resource Dictionary works AS LONG as it is defined after the Styles section. It was kind of a pain to translate the references but it works.

```xaml
<ResourceDictionary xmlns="https://github.com/avaloniaui"
                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">
    <SolidColorBrush x:Key="ButtonBackgroundApp" Color="{DynamicResource SystemAccentColorLight2}" Opacity=".35" />
    <StaticResource x:Key="ButtonBackground" ResourceKey="ButtonBackgroundApp" />

    <LinearGradientBrush x:Key="ButtonBorderBrushApp" StartPoint="0,0" EndPoint="0,3" Opacity=".35">
        <LinearGradientBrush.GradientStops>
            <GradientStop Offset="0.33" Color="{StaticResource ControlStrokeColorOnAccentSecondary}" />
            <GradientStop Offset="1.0" Color="{StaticResource ControlStrokeColorOnAccentDefault}" />
        </LinearGradientBrush.GradientStops>
    </LinearGradientBrush>
    <StaticResource x:Key="ButtonBorderBrush" ResourceKey="ButtonBorderBrushApp" />

    <SolidColorBrush x:Key="ButtonBackgroundPointerOverApp" Color="{DynamicResource SystemAccentColorLight2}"
                     Opacity=".30" />
    <StaticResource x:Key="ButtonBackgroundPointerOver" ResourceKey="ButtonBackgroundPointerOverApp" />

    <LinearGradientBrush x:Key="ButtonBorderBrushPointerOverApp" StartPoint="0,0" EndPoint="0,3" Opacity=".35">
        <LinearGradientBrush.GradientStops>
            <GradientStop Offset="0.33" Color="{StaticResource ControlStrokeColorOnAccentSecondary}" />
            <GradientStop Offset="1.0" Color="{StaticResource ControlStrokeColorOnAccentDefault}" />
        </LinearGradientBrush.GradientStops>
    </LinearGradientBrush>
    <StaticResource x:Key="ButtonBorderBrushPointerOver" ResourceKey="ButtonBorderBrushPointerOverApp" />

    <SolidColorBrush x:Key="ButtonBackgroundPressedApp" Color="{DynamicResource SystemAccentColorLight2}" Opacity=".27" />
    <StaticResource x:Key="ButtonBackgroundPressed" ResourceKey="ButtonBackgroundPressedApp" />

    <SolidColorBrush x:Key="ButtonBorderBrushPressedApp" Color="{DynamicResource ControlFillColorTransparent}"
                     Opacity=".35" />
    <StaticResource x:Key="ButtonBorderBrushPressed" ResourceKey="ButtonBorderBrushPressedApp" />

    <Color x:Key="WindowBackgroundGradiant">Green</Color>
    <LinearGradientBrush x:Key="WindowBackgroundBrush" EndPoint="50%,100%" StartPoint="50%,0%" Opacity=".25">
        <GradientStop Color="{DynamicResource Transparent}" Offset="0" />
        <GradientStop Color="{DynamicResource WindowBackgroundGradiant}" Offset="1" />
    </LinearGradientBrush>
</ResourceDictionary>
```

**Reply by timunie** [reply]
That's not suprising. When FluentAvalonia is applied, it will set it's resources, ignoring any further resources. So as @maxkatz6 said, the order of setting the resources can be important.

**Comment by mysteryx93** [comment]
Another question... what is the easiest way to have an SVG icon button match the foreground color?

**Reply by maxkatz6** [reply]
I don't think Svg.Skia avalonia control supports foreground inheritance of something like this.
Personally, if I need a SVG icon, I would have it with hardcoded color. Maybe duplicated.
But normally I prefer to replace SVG with XAML geometry which supports dynamic resource. Though not every SVG feature can be translated.

**Reply by maxkatz6** [reply]
I used this one to convert, but path syntax is compatible mostly https://github.com/wieslawsoltes/SvgToXaml

**Reply by timunie** [reply]
Can you open a new thread or ask this on telegram? I think this question is off topic

**Reply by maxkatz6** [reply]
I hope there is a better way to inherit foreground color for SVG

**Reply by mysteryx93** [reply]
Perhaps one option is to convert SVG into a font file; then I deal with normal text. I'll explore this tomorrow.
https://glyphter.com/

**Reply by timunie** [reply]
It's easy to copy & paste the path data from svg to xaml. You can edit the svg file in any notepad or in Inkscape and copy the data, or use the tool @maxkatz6 suggested.

**Reply by mysteryx93** [reply]
OK a Plus icon will translate to

```xaml
<Image>
  <Image.Resources>
    <SolidColorBrush x:Key="Brush0" Color="#FF000000" />
  </Image.Resources>
  <DrawingImage>
    <DrawingGroup>
      <GeometryDrawing Brush="Transparent" Geometry="F1M0,0L256,0L256,256L0,256z" />
      <DrawingGroup>
        <DrawingGroup.Transform>
          <MatrixTransform Matrix="0.72,0,0,0.72,128,128" />
        </DrawingGroup.Transform>
        <DrawingGroup>
          <DrawingGroup.Transform>
            <MatrixTransform Matrix="3.8899999,0,0,3.8899999,-175.05,-175.05" />
          </DrawingGroup.Transform>
          <GeometryDrawing Brush="{DynamicResource Brush0}" Geometry="F1 M45 90C42.239 90 40 87.762 40 85L40 5C40 2.239 42.239 0 45 0C47.762 0 50 2.239 50 5L50 85C50 87.762 47.762 90 45 90Z" />
          <GeometryDrawing Brush="{DynamicResource Brush0}" Geometry="F1 M85 50L5 50C2.239 50 0 47.762 0 45C0 42.239 2.239 40 5 40L85 40C87.762 40 90 42.239 90 45C90 47.762 87.762 50 85 50Z" />
        </DrawingGroup>
      </DrawingGroup>
    </DrawingGroup>
  </DrawingImage>
</Image>
```

I see 2 problems here.

First, how will I define this in a clean way to make it re-usable?

Second, I need to define the Brush; how can I take the Forecolor brush? The current style changes the forecolor on press.

Now let me explore the icon font option for drawbacks.

**Reply by mysteryx93** [reply]
The website for creating fonts does not allow saving products for free and thus the font we create cannot be edited afterwards. We can however use open source FontForge to easily create an icons font in TTF or OTF. Loaded it from app resource, shows up great with the desired forecolor! This is probably the best option.

**Reply by timunie** [reply]
Still the wrong thred. Anyway, this is one option. For others, I would just use PathIcon or simple Path which accepts Data. This works with scaling etc. No need for Image or other quirks. 

Note: in FluentAvalinia either use thier PathIcon or add the needed styles on your own.

**Reply by mysteryx93** [reply]
ah most icons I need are in SymbolIcon and work great!

FontIcon; I'm not sure what benefit it brings over this

    <Button Classes="icon" Width="35" Content="I" />

Ah I get it. IconSourceElement displays shareable icon from resources, which can be defined in a shared way like this

    <ui:PathIconSource x:Key="SamplePathIcon" Data="PATHDATA" />

That solves how to share path data.

Also I often need to run the FluentAvalonia sample project, which allows browsing all resources and colors. Great help!

**Reply by mysteryx93** [reply]
Looking at PathIcon, one icon translates as this using the app

```
<SolidColorBrush x:Key="Brush0" Color="#FF000000" />
<Image x:Key="InformationImage">
    <DrawingImage>
        <DrawingGroup>
            <DrawingGroup>
                <DrawingGroup.Transform>
                    <MatrixTransform Matrix="0.72,0,0,0.72,128,128" />
                </DrawingGroup.Transform>
                <DrawingGroup>
                    <DrawingGroup.Transform>
                        <MatrixTransform Matrix="3.8899999,0,0,3.8899999,-175.05,-175.05" />
                    </DrawingGroup.Transform>
                    <GeometryDrawing Brush="{DynamicResource Brush0}"
                                     Geometry="F1 M49.083 71.489L54.859 49.529L59.045 34.282C62.542 18.102 26.341 31.843 21.043 35.977L21.468 40.83C26.292 37.435 44.559 33.086 40.917 45.105L39.283 51.24L39.283 51.24L30.954 82.311C27.457 98.491 63.658 84.75 68.956 80.616L68.531 75.763C63.708 79.159 45.441 83.508 49.083 71.489Z" />
                    <GeometryDrawing Brush="{DynamicResource Brush0}"
                                     Geometry="F1 M65.072 11.201Q65.072 11.476 65.0585 11.7506Q65.045 12.0252 65.0181 12.2989Q64.9911 12.5725 64.9508 12.8445Q64.9104 13.1165 64.8568 13.3862Q64.8031 13.6559 64.7363 13.9226Q64.6695 14.1893 64.5897 14.4525Q64.5099 14.7156 64.4172 14.9745Q64.3246 15.2334 64.2194 15.4874Q64.1141 15.7415 63.9966 15.99Q63.879 16.2386 63.7494 16.4811Q63.6198 16.7236 63.4784 16.9595Q63.337 17.1953 63.1843 17.4239Q63.0315 17.6526 62.8677 17.8734Q62.7039 18.0943 62.5295 18.3068Q62.355 18.5194 62.1704 18.7231Q61.9857 18.9269 61.7913 19.1213Q61.5969 19.3157 61.3931 19.5004Q61.1894 19.6851 60.9768 19.8595Q60.7643 20.0339 60.5434 20.1977Q60.3226 20.3615 60.0939 20.5143Q59.8653 20.6671 59.6295 20.8084Q59.3936 20.9498 59.1511 21.0794Q58.9086 21.209 58.66 21.3266Q58.4115 21.4441 58.1574 21.5494Q57.9034 21.6546 57.6445 21.7472Q57.3856 21.8399 57.1225 21.9197Q56.8593 21.9995 56.5926 22.0663Q56.3259 22.1331 56.0562 22.1868Q55.7865 22.2404 55.5145 22.2808Q55.2425 22.3211 54.9689 22.3481Q54.6952 22.375 54.4206 22.3885Q54.146 22.402 53.871 22.402Q53.596 22.402 53.3214 22.3885Q53.0467 22.375 52.7731 22.3481Q52.4995 22.3211 52.2275 22.2808Q51.9555 22.2404 51.6858 22.1868Q51.4161 22.1331 51.1494 22.0663Q50.8826 21.9995 50.6195 21.9197Q50.3564 21.8399 50.0975 21.7472Q49.8386 21.6546 49.5846 21.5494Q49.3305 21.4441 49.0819 21.3266Q48.8334 21.209 48.5909 21.0794Q48.3484 20.9498 48.1125 20.8084Q47.8767 20.6671 47.648 20.5143Q47.4194 20.3615 47.1986 20.1977Q46.9777 20.0339 46.7652 19.8595Q46.5526 19.6851 46.3489 19.5004Q46.1451 19.3157 45.9507 19.1213Q45.7563 18.9269 45.5716 18.7231Q45.3869 18.5194 45.2125 18.3068Q45.0381 18.0943 44.8743 17.8734Q44.7105 17.6526 44.5577 17.4239Q44.4049 17.1953 44.2636 16.9595Q44.1222 16.7236 43.9926 16.4811Q43.863 16.2386 43.7454 15.99Q43.6278 15.7415 43.5226 15.4874Q43.4174 15.2334 43.3248 14.9745Q43.2321 14.7156 43.1523 14.4525Q43.0725 14.1893 43.0057 13.9226Q42.9389 13.6559 42.8852 13.3862Q42.8316 13.1165 42.7912 12.8445Q42.7509 12.5725 42.7239 12.2989Q42.697 12.0252 42.6835 11.7506Q42.67 11.476 42.67 11.201Q42.67 10.926 42.6835 10.6514Q42.697 10.3768 42.7239 10.1031Q42.7509 9.82946 42.7912 9.55747Q42.8316 9.28548 42.8852 9.01579Q42.9389 8.74611 43.0057 8.47938Q43.0725 8.21265 43.1523 7.94952Q43.2321 7.68639 43.3248 7.4275Q43.4174 7.1686 43.5226 6.91456Q43.6278 6.66052 43.7454 6.41195Q43.863 6.16339 43.9926 5.92089Q44.1222 5.67838 44.2636 5.44253Q44.4049 5.20669 44.5577 4.97806Q44.7105 4.74943 44.8743 4.52857Q45.0381 4.30771 45.2125 4.09516Q45.3869 3.88261 45.5716 3.67887Q45.7563 3.47513 45.9507 3.2807Q46.1451 3.08626 46.3489 2.90161Q46.5526 2.71695 46.7652 2.54251Q46.9777 2.36807 47.1986 2.20427Q47.4194 2.04047 47.648 1.88771Q47.8767 1.73494 48.1125 1.59358Q48.3484 1.45222 48.5909 1.3226Q48.8334 1.19298 49.0819 1.07542Q49.3305 0.957851 49.5846 0.852625Q49.8386 0.747399 50.0975 0.654765Q50.3564 0.562131 50.6195 0.482311Q50.8826 0.402492 51.1494 0.33568Q51.4161 0.268868 51.6858 0.215224Q51.9555 0.16158 52.2275 0.121234Q52.4995 0.0808875 52.7731 0.0539359Q53.0467 0.0269842 53.3214 0.0134921Q53.596 0 53.871 0Q54.146 0 54.4206 0.0134921Q54.6952 0.0269842 54.9689 0.0539359Q55.2425 0.0808875 55.5145 0.121234Q55.7865 0.16158 56.0562 0.215224Q56.3259 0.268868 56.5926 0.33568Q56.8593 0.402492 57.1225 0.482311Q57.3856 0.562131 57.6445 0.654765Q57.9034 0.747399 58.1574 0.852625Q58.4115 0.957851 58.66 1.07542Q58.9086 1.19298 59.1511 1.3226Q59.3936 1.45222 59.6295 1.59358Q59.8653 1.73494 60.0939 1.88771Q60.3226 2.04047 60.5434 2.20427Q60.7643 2.36807 60.9768 2.54251Q61.1894 2.71695 61.3931 2.90161Q61.5969 3.08626 61.7913 3.2807Q61.9857 3.47513 62.1704 3.67887Q62.355 3.88261 62.5295 4.09516Q62.7039 4.30771 62.8677 4.52857Q63.0315 4.74943 63.1843 4.97806Q63.337 5.20669 63.4784 5.44253Q63.6198 5.67838 63.7494 5.92088Q63.879 6.16339 63.9966 6.41195Q64.1141 6.66052 64.2194 6.91456Q64.3246 7.1686 64.4172 7.4275Q64.5099 7.68639 64.5897 7.94952Q64.6695 8.21265 64.7363 8.47938Q64.8031 8.74611 64.8568 9.01579Q64.9104 9.28548 64.9508 9.55747Q64.9911 9.82946 65.0181 10.1031Q65.045 10.3768 65.0585 10.6514Q65.072 10.926 65.072 11.201Z" />
                </DrawingGroup>
            </DrawingGroup>
        </DrawingGroup>
    </DrawingImage>
</Image>
```

1. This can not be translated into a simple PATH string, so I cannot use PathIcon
2. ImageIcon is not working with this either; thus I can't turn it into any IconSource
3. I need to set the brush manually; so if the Foreground color changes on press, that won't reflect it.

Which brings me back to FontIcon being the only option?

This works though; except for the Forecolor.

    <Button Content="{StaticResource InformationImage}" />

**Reply by maxkatz6** [reply]
Does ImageIcon support DrawingImage? If so, you can translate it to XAML.

**Reply by timunie** [reply]
ImageIcon is from FluentAvalonia I think. 

FYI: In Inkscape you can merge several pathes and also you can analyze the svg nodes. See: 
- https://inkscape.org
- https://goinkscape.com/how-to-merge-objects-in-inkscape/ 
- https://graphicdesign.stackexchange.com/questions/147506/inkscape-to-convert-multiple-paths-into-one-single-path

After that you can either read the svg in Notepad or use Inkscape build in XML viewer and copy path data.

**Reply by mysteryx93** [reply]
Great I'm learning a lot about SVG...  after installing Inkscape, merging it into a single path, converting it to XAML with the other path. It still displays with completely the wrong size making it unusable.

Probably that can be fixed, but...

I could also argue that importing the icon into a font file is easier!

**Reply by maxkatz6** [reply]
If it's a monochrome font, then yeah, why not. It would be easier to manage this way probably.

**Reply by timunie** [reply]
> Great I'm learning a lot about SVG...  after installing Inkscape, merging it into a single path, converting it to XAML with the other path. It still displays with completely the wrong size making it unusable.
> 
> Probably that can be fixed, but...
> 
> I could also argue that importing the icon into a font file is easier!

Yes, sure. If a font works for you, go that route. I think it's always good to know different options 👍. Now you have at least two 😉

**Comment by mysteryx93** [comment]
Here's a weird issue.

I got this window background brush; applies a semi-transparent  gradual tint over the background, whether white or dark.

```xaml
<Color x:Key="WindowBackgroundGradiant">Green</Color>
<LinearGradientBrush x:Key="WindowBackgroundBrush" EndPoint="50%,100%" StartPoint="50%,0%" Opacity=".25">
    <GradientStop Color="{DynamicResource Transparent}" Offset="0" />
    <GradientStop Color="{DynamicResource WindowBackgroundGradiant}" Offset="1" />
</LinearGradientBrush>
```

If I set this, switching theme works.

    <Grid Background="{DynamicResource WindowBackgroundBrush}">

But if I set the background directly on the Window, then the background doesn't switch! Without that tint brush, the window background switches without problem.

Why?

**Reply by maxkatz6** [reply]
Hard to tell without simple example.
Might be related to a bug, that app seem to not invalidate when gradient stop color was changed https://github.com/AvaloniaUI/Avalonia/issues/8266
